### PR TITLE
Master mdl 71526 34

### PR DESCRIPTION
**Pull request by safatshahin**
PR for #34

### CODE REVIEW

**Review by guillogo — COMMENTED** [review]
Hi @safatshahin 

I've finished the review:

1. Tested and working as expected.
2. Added comment to commits.

Can you have a look to these please:

1. My review comments on commits.
2. For the new files on `questions/local/bank/` can you please check that all function follow this code style: `public function sampleFunction(int $a, int $b = null): array` (except the parameter part for the ones that are overriding functions).
3. Re-check the questions/local/bank/view.php please, I'm getting this in my CodeChecker: 

```
#220: ················if(!$classobject->get_preference())·{
Expected "if (...) {\n"; found "if(...) {\n"
#384: ················//'core_question\local\bank\question_name_idnumber_tags_column-name'·=>·1
Inline comments must start with a capital letter, digit or 3-dots sequence
No space found before comment text; expected "// 'core_question\local\bank\question_name_idnumber_tags_column-name' => 1" but found "//'core_question\local\bank\question_name_idnumber_tags_column-name' => 1"
#1016: ················print_error('cannotfindcate',·'question');
Function print_error() has been deprecated; use throw new moodle_exception() (using lang strings only if meant to be shown to final user)() instead
#1063: ····················print_error('invalidconfirm',·'question');
Function print_error() has been deprecated; use throw new moodle_exception() (using lang strings only if meant to be shown to final user)() instead
#1085: ············$questionlist·=·'';··//·comma·separated·list·of·ids·of·questions·to·be·deleted
Inline comments must start with a capital letter, digit or 3-dots sequence
Inline comments must end in full-stops, exclamation marks, or question marks
#1086: ············$questionnames·=·'';·//·string·with·names·of·questions·separated·by·<br·/>·with
Inline comments must start with a capital letter, digit or 3-dots sequence
#1087: ············//·an·asterix·in·front·of·those·that·are·in·use
Inline comments must end in full-stops, exclamation marks, or question marks
#1088: ············$inuse·=·false;······//·set·to·true·if·at·least·one·of·the·questions·is·in·use
Inline comments must start with a capital letter, digit or 3-dots sequence
Inline comments must end in full-stops, exclamation marks, or question marks
```

Thank you.
**Guillermo**

**Comment by safatshahin** [comment]
Hi @guillogo 
thanks for reviewing this. i have made the changes according to the comments.
some notes:
i had to revert the changes to the classes in /bank which were pointing to the /local/bank, as i added return types, those child classes were throwing errors as they had mismatch in return types. if we are to keep those, we have to fix all the deprecated child classes.
please have a look now.
thanks

File: question/engine/lib.php [hunk 1]
```diff
@@ -51,6 +51,7 @@
  * which is the main class that is used by other code that wants to use questions.
  *
  * @copyright  2009 The Open University
+ * @author     Safat Shahin <safatshahin@catalyst-au.net>
```
Comment (guillogo) [inline question/engine/lib.php:54]: can you please delete this line as we aren't editing this one

File: question/classes/local/bank/view.php [hunk 2]
```diff
@@ -0,0 +1,1191 @@
...
+
+    /**
+     * Default sort for question data.
+     * @return int[]
+     */
+    protected function default_sort(): array {
+        // Change required after implementing as plugins.
+        return array(
+                'core_question\local\bank\question_type_column' => 1,
+                //'core_question\local\bank\question_name_idnumber_tags_column-name' => 1
```
Comment (guillogo) [inline question/classes/local/bank/view.php:401]: This comment will fail in the codecheceker. should be something like

``// ...'core_question\local\bank\question_name_idnumber_tags_column-name' => 1``

Space after // and 2 dots if we want to start with lowercase.

File: question/classes/local/bank/column_base.php [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        return implode(' ', $classes);
+    }
+
+    /**
+     * Get the internal name for this column. Used as a CSS class name,
+     * and to store information about the current sort. Must match PARAM_ALPHA.
+     *
+     * @return string column name.
+     */
+    abstract public function get_name(): string;
```
Comment (guillogo) [inline question/classes/local/bank/column_base.php:229]: abstract methods doesn't need the return.

File: question/classes/local/bank/column_base.php [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    public function get_extra_classes(): array {
+        return array();
+    }
+
+    /**
+     * Output the contents of this column.
+     * @param object $question the row from the $question table, augmented with extra information.
+     * @param string $rowclasses CSS class names that should be applied to this row of output.
+     */
+    abstract protected function display_content($question, $rowclasses): void;
```
Comment (guillogo) [inline question/classes/local/bank/column_base.php:243]: abstract methods doesn't need the return.

**Review by guillogo — COMMENTED** [review]
Hi @safatshahin 

Added comments for column_base and view.

Thank you.
**Guillermo**

**Comment by safatshahin** [comment]
Hi @guillogo 
Resolved the comments, please have a look now.
thanks

**Pull request merged**